---
title: "R Notebook"
output: html_notebook
---

#Importing libraries
Importing dada2, phyloseq, ggplot2, seqinr, decodam, DESeq2, Biostrings, microeco, ape, ShortRead, dendextend, dplyr, ggpubr, fantaxtic, ggforce, microbiomeutilities, pheatmap, vegan, microbiome

```{r}
library('dada2')
library("phyloseq")
library("seqinr")
library("decontam")
library("ggplot2")
library("Biostrings")
library("DESeq2")
library("microeco")
library("ape")
library(ShortRead)
library(dendextend)
library(dplyr)
library(ggpubr)
library(fantaxtic)
library(ggforce)
library(microbiomeutilities)
library(pheatmap)
library(vegan)
library(microbiome)
```

Set theme for all ggplots 

```{r}
theme_set(theme_bw())
```

##Making files for analysis
#Import and filter files
CHANGE the path to the directory containing the fastq files after unzipping.
```{r}
path <- "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/MiSeq_SOP" 
list.files(path)
```

Take filenames from directory and make list of it
```{r}
fnFs <- sort(list.files(path, pattern=".R1.fastq", full.names = TRUE)) #достаем имена файлов
fnRs <- sort(list.files(path, pattern=".R2.fastq", full.names = TRUE))

sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

Plot quality of seq
```{r}
plotQualityProfile(fnFs[1:10])
plotQualityProfile(fnRs[1:10])
```

Filter files and trunclen of our 16sRNA with normal quality of sample
```{r}
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))

names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(130,130),
                     maxN=0, maxEE=Inf, truncQ=2, rm.phix=FALSE,
                     compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

#Errors
Learn errors in our files, print plots and clean our data
```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)

dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
```

#Merge files
Merge files and inspect the merger data.frame from the first sample (Merge[1] shows exampl of mergers.rep). After it construct seqtab table with our seq's with counts of it in every bee.
```{r}
mergers.rep <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, minOverlap=9, verbose=TRUE)
mergers.rep[[1]]

seqtab <- makeSequenceTable(mergers.rep)
dim(seqtab)
table(nchar(getSequences(seqtab)))
```
Remove chimeras and analys our data after it. If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs).
```{r}
seqtab.nochim.rep <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim.rep)
sum(seqtab.nochim.rep)/sum(seqtab)

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers.rep, getN), rowSums(seqtab.nochim.rep))

colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```
```{r}
write.csv(seqtab.nochim.rep,"/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/seqtab.nochim.rep.csv", row.names = FALSE)
seqtab.nochim.rep1 <- seqtab.nochim.rep
#saveRDS(seqtab, "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/seqtab.rep.rds") 
```

#Assign Taxonomy
Assign taxonomy and print it. Take data of taxas from directory

```{r}

taxa.rep <- assignTaxonomy(seqtab.nochim.rep, "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE)
taxa.rep <- addSpecies(taxa.rep, "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/silva_species_assignment_v138.1.fa.gz")
taxa.print <- taxa.rep # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

```

Save sequences as fasta to the current directory. Here we should change directory for files from diffrent data! 
```{r}
asvnames <- sprintf("ASV_%04d", seq(1,nrow(taxa.rep)))
write.fasta(as.list(rownames(taxa.rep)), asvnames, 
            file.out = "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/asvs.rep.fasta",  open = "w", as.string = TRUE)

rownames(taxa.rep) <- asvnames
colnames(seqtab.nochim.rep) <- asvnames
```

#Making first phyloseq oject metadata
Connect Silva taxonomy to phyloseq (4 export only) and making count table with asv's rows and bee's columns
```{r}
ps.Silva.rep <- phyloseq(otu_table(seqtab.nochim.rep, taxa_are_rows=FALSE),
                     sample_data(sample.metadata.rep),
                     tax_table(taxa.rep))

readcount.table <- cbind(t(ps.Silva.rep@otu_table),ps.Silva.rep@tax_table)
```

#importing files
Importing files to thw directory. Change directoty for next analysis */Users/aleksandrakozlova/Desktop/Work/Kurchatnik/*
```{r}
write.csv(ps.Silva.rep@tax_table, file = "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/dada2.Silva.taxtab.rep.csv")
write.csv(t(ps.Silva.rep@otu_table), file = "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/dada2.Silva.otutab.rep.csv")
write.csv(readcount.table, file = "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/dada2.Silva.sumtab.rep.csv")
```



##Import files that are ready for analysis
#Importing metadata 
Importing metadata from current dirrectory (header=FALSE). After it change the names of columns. Directory start = */Users/aleksandrakozlova/Desktop/Work/Kurchatnik/*.
```{r}
sample.metadata.rep <- sample_data(read.csv(file = '/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/MiSeq_SOP/meta_data - Copy of Sheet1.csv',
                                        dec = ",",
                                        header = TRUE,
                                        row.names = 1 ) )
```

Import libraries back to the file. Change directoty for next analysis */Users/aleksandrakozlova/Desktop/Work/Kurchatnik/*
```{r}
otu <- otu_table(as.matrix(read.csv(file = "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/MiSeq_SOP/dada2.Silva.otutab.rep.csv", row.names = 1)), taxa_are_rows = TRUE)
tax <- tax_table(as.matrix(read.csv(file = "/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/MiSeq_SOP/dada2.Silva.taxtab.rep.csv", row.names = 1)))
```

#Make phyloseq for analysis 
Make phyloseq for analysis and removing mitochondria and c
```{r}
fin.phy <- phyloseq(sam_data = sample.metadata.rep, otu_table = otu, tax_table = tax, refseq = readcount.table)
fin.phy <- fin.phy %>% subset_taxa( Family!= "Mitochondria" | is.na(Family) & Class!="Chloroplast" | is.na(Class) )
```

##Analysis

#Make pheatmap

Firstly we should modify data for the next analysis.
Take genus for our asv's and make list. Take core members (with diffrent prevalence and detection). After that makeconnection with our phyloseq 

For genuses (*with different prevalence!*)
```{r}
tax_glom_fin.phy <- tax_glom(fin.phy, "Genus" )

taxas1 <- core_members(tax_glom_fin.phy, detection = 3, prevalence = 20/100)
ps_obj1 <- prune_taxa(taxas1, tax_glom_fin.phy)
```

For asvs
```{r}
taxas1 <- core_members(fin.phy, detection = 3, prevalence = 5/100)
ps_obj1 <- prune_taxa(taxas1, fin.phy)
```


Make two functions for make df from our phyloseq and connect data about taxa for our asv's.
```{r}
taxa_name_func <- function(x) {paste0(x['ASV'], '__', x['Family'],  '.g__' , gsub('/', '', gsub("-", "", x['Genus'])))}
prepare_count_table <- function (ps_obj, taxa_name_func) {
  otu_matrix <- as(otu_table(ps_obj), 'matrix')
  taxa_matrix <- as(tax_table(ps_obj), 'matrix')
  taxa_matrix <- cbind(ASV=rownames(taxa_matrix), taxa_matrix)  
  taxa_matrix_good_names <- apply(taxa_matrix, MARGIN=1, taxa_name_func)
  rownames(otu_matrix) <- taxa_matrix_good_names
  return(otu_matrix)
}
```

Make our function work and give some arguments. After that normalize our data with log or clr function.
```{r}
count_mtrx <- prepare_count_table(ps_obj1, taxa_name_func) #задаем то, что берется в функцию
ps2 <- microbiome::transform(count_mtrx, "clr")
```


Here we should count our data for the correlation data and other! 
```{r}

```

Function for saving pdf with pheatmap
```{r}
save_pheatmap_pdf <- function(x, filename, width=12, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
```

Connect metadata with df (don't know how take the exist metadata ;(((( )
```{r}
readfile <- read.csv('/Users/aleksandrakozlova/Desktop/Work/Kurchatnik/Ossetian.bees/meta_data - Copy of Sheet1.csv', header = TRUE)
readfile1 <- select(readfile, -Sample1, -bee_file, -Sample)
row.names(readfile1) <- colnames(ps2)

```

After all modifications we can plot the pheatmap and ssave it.
```{r}
xx <- pheatmap(ps2,cellwidth = 6, cellheight = 2.5,
         cluster_row = T,
         show_rownames = T,
         show_colnames = T,
         fontsize_row = 2, 
         fontsize_col = 5,
         fontsize = 5,
         annotation_col = readfile1, cluster_cols = FALSE, gaps_col = seq(2,82,by=2))

```

```{r}
xx$tree_row$order <- 
```

Save pheatmap for the dirrectory  *Change names in save_pheatmap_pdf!!!*
```{r}
save_pheatmap_pdf(xx, "clr_asv_5prc.pdf")
```

#Plot ordination

Transformation (otu/sum otu)
```{r}
ps.prop <- transform_sample_counts(fin.phy, function(otu) otu/sum(otu))
```
Drow ordination
```{r}
ps.prop <- transform_sample_counts(fin.phy, function(otu) otu/sum(otu))
ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray")

plot_ordination(fin.phy, ord.nmds.bray, color='Place', shape = 'bee.gut', title="Bray NMDS",label="Sample1")
```


#Abundance
Abundance of bacteria in honeybees
```{r}
boxplot(t(ps2)) +
  theme(axis.text.x = element_text(angle = 45))

#qplot( data = t(ps2), geom = "boxplot")

```



